adapters/vidoomy/vidoomy.go

```diff
@@ -0,0 +1,192 @@
+	headers := http.Header{}
+
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("x-openrtb-version", "2.5")
+
+	if request.Device == nil {
+		return headers
+	}
+
+	if request.Device.UA != "" {
+		headers.Set("User-Agent", request.Device.UA)
+	}
+
+	if len(request.Device.IPv6) > 0 {
+		headers.Add("X-Forwarded-For", request.Device.IPv6)
+	}
+
+	if len(request.Device.IP) > 0 {
+		headers.Add("X-Forwarded-For", request.Device.IP)
+	}
+
+	return headers
+}
+
+func changeRequestForBidService(request *openrtb2.BidRequest, extension *openrtb_ext.ImpExtVidoomy) error {
+	if request.Imp[0].Banner == nil {
+		return nil
+	}
+
+	banner := *request.Imp[0].Banner
+	request.Imp[0].Banner = &banner
+
+	if banner.W != nil && banner.H != nil {
+		if *banner.W == 0 || *banner.H == 0 {
+			return fmt.Errorf("invalid sizes provided for Banner %d x %d", *banner.W, *banner.H)
+		}
+		return nil
+	}
+
+	if len(banner.Format) == 0 {
+		return fmt.Errorf("no sizes provided for Banner %v", banner.Format)
+	}
+
+	banner.W = openrtb2.Int64Ptr(banner.Format[0].W)
+	banner.H = openrtb2.Int64Ptr(banner.Format[0].H)
+
+	return nil
+}
+
+func (a adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d.", response.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb2.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Bad server response: %d.", err),
+		}}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(internalRequest.Imp))
+
+	for _, sb := range bidResp.SeatBid {
+		for _, bid := range sb.Bid {
+			bid := bid
+			exists, mediaType := getImpInfo(bid.ImpID, internalRequest.Imp)
+			if !exists {
+				return nil, []error{&errortypes.BadServerResponse{
+					Message: fmt.Sprintf("Unknown ad unit code '%s'", bid.ImpID),
+				}}
+			}
+
+			if openrtb_ext.BidTypeBanner != mediaType &&
+				openrtb_ext.BidTypeVideo != mediaType {
+				//only banner and video are supported, anything else is ignored
+				continue
+			}
+
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: mediaType,
+			})
+		}
+	}
+
+	return bidResponse, nil
+}
+
+func getImpInfo(impId string, imps []openrtb2.Imp) (bool, openrtb_ext.BidType) {
+	var mediaType openrtb_ext.BidType
+	var exists bool
+	for _, imp := range imps {
+		if imp.ID == impId {
+			exists = true
+			if imp.Banner != nil {
+				mediaType = openrtb_ext.BidTypeBanner
+			} else if imp.Video != nil {
+				mediaType = openrtb_ext.BidTypeVideo
+			}
```

Comment:
Does this match the evaluation order of your bidding server? If an impression has both a banner and video format, is banner always bid for first?

Reply:
We support one impression per request so we have no order, just trying to get the impression type.

Reply:
Is there a way to determine from your bid response if the bid is for banner or video? If the one impression defines both options, what gives you confidence here that the correct impression type is selected?

Reply:
Good point!, Based on Openrtb 2.5 which says

`The presence of Banner (Section 3.2.6), Video (Section 3.2.7), and/or Native (Section 3.2.9) objects
subordinate to the Imp object indicates the type of impression being offered. The publisher can choose
one such type which is the typical case or mix them at their discretion. However, any given bid for the
impression must conform to one of the offered types.`

So my understanding is it is not possible to have an impression with multiple types!

Reply:
The `publisher can choose ... [a] mix ... at their discretion`, so the publisher may define an impression with both banner and video objects if they like. A bid is only for one., either banner or video in this case. The intent of this method is for the adapter to tell Prebid Server which format the bid is for. Some bidders include the bid type in their server response, some re-implement the same bidding logic in their adapter that their server follows, and some effectively choose one at random which could result in rejected bids.

I personally view this as an issue with the OpenRTB 2.5 spec itself. Luckily, OpenRTB 2.6 will fix this issue by including the media type (or "mtype") in the standard bid response. Until then, each adapter needs to do the best they can to match up the bid with the format. I just want to be sure that's the case here - that this is the best logic possible for Vidoomy to give the adapter the best shot at success.

Reply:
Thanks for your explanation and your effort, I can assure you that this is the best solution for us, we only support video and banner, if we receive an impression with both types available we chose Video over Banner so only one type will be available in impression.

Reply:
> if we receive an impression with both types available we chose Video over Banner

Cool. Could you please update the adapter code to match? Right now, it will prefer banner over video. If you put the video check first, then you'll have matching behavior:

```
if imp.Video != nil {
	mediaType = openrtb_ext.BidTypeVideo
} else if imp.Banner != nil {
	mediaType = openrtb_ext.BidTypeBanner
}
```

Reply:
Since it is not possible to have multiple types based on my explanation, yet I changed the order.